lib/class-wp-rest-batch-controller.php

```diff
@@ -0,0 +1,298 @@
+						array( 'status' => 400 )
+					);
+				}
+			}
+
+			if ( ! $error ) {
+				$check_required = $single_request->has_valid_params();
+				if ( is_wp_error( $check_required ) ) {
+					$error = $check_required;
+				}
+			}
+
+			if ( ! $error ) {
+				$check_sanitized = $single_request->sanitize_params();
+				if ( is_wp_error( $check_sanitized ) ) {
+					$error = $check_sanitized;
+				}
+			}
+
+			if ( $error ) {
+				$has_error    = true;
+				$validation[] = $error;
+			} else {
+				$validation[] = true;
+			}
+		}
+
+		$responses = array();
+
+		if ( $has_error && 'require-all-validate' === $batch_request['validation'] ) {
+			foreach ( $validation as $valid ) {
+				if ( is_wp_error( $valid ) ) {
+					$responses[] = rest_get_server()->envelope_response( $this->error_to_response( $valid ), false )->get_data();
+				} else {
+					$responses[] = null;
+				}
+			}
+
+			return new WP_REST_Response(
+				array(
+					'failed'    => 'validation',
+					'responses' => $responses,
+				),
+				WP_Http::MULTI_STATUS
+			);
+		}
+
+		foreach ( $requests as $i => $batch_request ) {
+			$clean_request = clone $batch_request;
+			$clean_request->set_url_params( array() );
+			$clean_request->set_attributes( array() );
+			$clean_request->set_default_params( array() );
+
+			/** This filter is documented in wp-includes/rest-api/class-wp-rest-server.php */
+			$result = apply_filters( 'rest_pre_dispatch', null, rest_get_server(), $clean_request );
+
+			if ( empty( $result ) ) {
+				$match = $matches[ $i ];
+				$error = null;
+
+				if ( is_wp_error( $validation[ $i ] ) ) {
+					$error = $validation[ $i ];
+				}
+
+				if ( is_wp_error( $match ) ) {
+					$result = $this->error_to_response( $match );
+				} else {
+					list( $route, $handler ) = $match;
+
+					if ( ! $error && ! is_callable( $handler['callback'] ) ) {
+						$error = new WP_Error(
+							'rest_invalid_handler',
+							__( 'The handler for the route is invalid', 'gutenberg' ),
+							array( 'status' => 500 )
+						);
+					}
+
+					$result = rest_get_server()->respond_to_request( $batch_request, $route, $handler, $error );
+				}
+			}
+
+			/** This filter is documented in wp-includes/rest-api/class-wp-rest-server.php */
+			$result = apply_filters( 'rest_post_dispatch', rest_ensure_response( $result ), rest_get_server(), $batch_request );
+
+			$responses[] = rest_get_server()->envelope_response( $result, false )->get_data();
+		}
+
+		return new WP_REST_Response( array( 'responses' => $responses ), WP_Http::MULTI_STATUS );
+	}
+
+	/**
+	 * Polyfills a simple form of batching for compatibility for non-trunk installs.
+	 *
+	 * @since 9.0.0
+	 *
+	 * @param WP_REST_Request[] $requests The list of requests to perform.
+	 * @return WP_REST_Response The response object.
+	 */
+	protected function polyfill_batching( $requests ) {
+		$responses = array();
+
+		foreach ( $requests as $request ) {
+			if ( 0 !== strpos( $request->get_route(), '/__experimental' ) ) {
```

Comment:
Explicitly marking each route as eligible/ineligible for batch requests seems cleaner. Maybe this method will be gone soon enough that this check suffices?

Reply:
Ah it's already there in `$handler['allow_batch']`, it's just we can't make use of it here because it's a polyfill. Makes sense!

Reply:
Yep, this is just a fallback since there isn't a good way to get that `allow_batch` flag in the polyfilled version.